### DISCUSSION TITLE
USBSerial induces extra power consumption in deep sleep

### DISCUSSION THREAD

**Discussion by AlexandreCampo** [stm32duino/Arduino_Core_STM32#1718] [Q&A]
Hi, 
I am working with a custom board with an STM32L476RG, and trying to get low power in deep sleep.

Before entering deep sleep with STM32LowPower library, I put some peripherals in reset:

```
                __HAL_RCC_ADC_FORCE_RESET();
                __HAL_RCC_I2C1_FORCE_RESET();
                __HAL_RCC_I2C2_FORCE_RESET();
                __HAL_RCC_I2C3_FORCE_RESET();
                __HAL_RCC_USART1_FORCE_RESET();
                __HAL_RCC_USART2_FORCE_RESET();
                __HAL_RCC_USART3_FORCE_RESET();
                __HAL_RCC_UART4_FORCE_RESET();
                __HAL_RCC_UART5_FORCE_RESET();
```

If USB CDC is NOT enabled, this gives me a 1.84uA consumption in deep sleep. However, if USB CDC is enabled and USBSerial
initialized, consumption climbs to 608uA.

I tried:
```
                Serial.end();
                __HAL_RCC_USB_OTG_FS_FORCE_RESET();
```
Which leads to 404uA in deep sleep.

I was hoping to reach the nice 1.84uA, but I am stuck here...
Is there any way to disable USB in deep sleep as if it was never initialized and remove any extra consumption?

Thanks for your help!

**Comment by fpistm** [comment]
Hi @AlexandreCampo 
Maybe you can try to end() the SerialUSB interface. 
One other way would be to enable the low power at USB level but I never try it:
https://github.com/stm32duino/Arduino_Core_STM32/blob/0e00a8d8904d867acba88ddb313a804d4670b0b0/cores/arduino/stm32/usb/usbd_conf.c#L470

**Reply by AlexandreCampo** [reply]
Hi Frederic,  thanks for your quick feedback!

I tried SerialUSB.end() (actually Serial.end() since it is an alias) but it does leaves some extra consumption, as mentioned in my original post.

I further checked and tested enabling low_power_enable flag in usbd_conf.c but this also has no noticeable effect. 
From what I understand of that part of the code, it seems more aimed at handling USB wake up in low power mode.

I also tested separately and without success:
```
                 __HAL_RCC_PWR_CLK_DISABLE();
                 __HAL_RCC_SYSCFG_CLK_DISABLE();
                __HAL_PCD_GATE_PHYCLOCK(&g_hpcd);
                __HAL_PCD_DISABLE(&g_hpcd);
```

If you happen to have any other suggestion I would be interested. 
I plan to submit a feature request to be able to turn USB off entirely, in the hope that maintainers of that code can help.

**Reply by fpistm** [reply]
Currently, I have no more idea.
Here the description in the USB Device library user manual:
![image](https://user-images.githubusercontent.com/20641798/169081389-30a4f7a0-50f0-4a20-91d6-74d5e8acbd9e.png)

You can submit a feature request, it will be added to https://github.com/stm32duino/Arduino_Core_STM32/issues/687
Any contribution are welcome. 😉

**Reply by AlexandreCampo** [reply]
Ok, thanks for your responsivness! Would be happy to help if I figure out what to do :-)

**Comment by cristhianguedes** [comment]
Hi Alexandre.
Did you get any resolution for this?
I'm having the same problem.
